File: galton_board_experiment/simulation.js

n=14 // number of experiments (go left or go right)
size=10 // size of a grid cell
startNbRows=4
endNbRows=50
nbRows=startNbRows+2*n+endNbRows
nbCols=n*4+1
canW=nbCols*size
canH=nbRows*size
listBalls=[]
totalBalls=600
countBall=0
grid=[]
centerC=Math.floor(nbCols/2)
durationIntervalAdd=0.3
durationIntervalMove=durationIntervalAdd/2
timerIntervalAdd=durationIntervalAdd
timerIntervalMove=durationIntervalMove

oldTime=0

window.onload=initialisation
function initialisation(){
    can=document.getElementById("can")
    can.width=canW
    can.height=canH
    ctx=can.getContext("2d")
    initGrid()
    addBall()
    requestAnimationFrame(gameLoop)
}

function initGrid(){
    for(let i=0; i<nbRows; i++){
        grid[i]=[]
        for(let j=0; j<nbCols; j++){
            if (i<startNbRows-1){
                if (j==centerC-1){
                    grid[i][j]=3
                }else if(j==centerC+1){
                    grid[i][j]=4
                }else{
                    grid[i][j]=0
                }
            }else if (i<startNbRows){
                grid[i][j]=0
            }else if (i<startNbRows+2*n){
                if ((i-startNbRows)%4==0){
                    if(j%4==0){
                        grid[i][j]=1
                    }else{
                        grid[i][j]=0
                    }
                }else if ((i-startNbRows)%4==1 || (i-startNbRows)%4==3){
                    grid[i][j]=0
                }else if ((i-startNbRows)%4==2){
                    if (j%4==2){
                        grid[i][j]=1
                    }else{
                        grid[i][j]=0
                    }
                }
            }else{
                if(j%4==2){
                    grid[i][j]=2
                }else{
                    grid[i][j]=0
                }
            }
        }    
    }
}
function getRandomInt(nb){
    return Math.floor(Math.random()*nb)
}

function addBall(){
    let ball={}
    ball.x=centerC*size+size/2
    ball.y=size/2
    ball.r=size/2
    ball.vx=0
    ball.vy=size
    ball.color=[getRandomInt(256), getRandomInt(256), getRandomInt(256)]
    listBalls.push(ball)
}

function update(dt){
    timerIntervalAdd-=dt 
    timerIntervalMove-=dt
    //console.log(timerIntervalAdd, timerIntervalMove)
    if (timerIntervalMove<=0){
        updateBalls()
        timerIntervalMove=durationIntervalMove
    } 
    if (countBall < totalBalls){
        if (timerIntervalAdd<=0){
            addBall()
            countBall++
            timerIntervalAdd=durationIntervalAdd
        }
    } 
}

function updateBalls(){
    for(let i=0; i<listBalls.length; i++){
        let ball=listBalls[i]
        ball.x=ball.x+ball.vx
        ball.y=ball.y+ball.vy
        let r=Math.floor((ball.y-size/2)/size)
        let c=Math.floor((ball.x-size/2)/size)
        let belowBall=below(r,c)
        if(r<startNbRows+2*n-1){
            if(belowBall==1){
                let num=Math.random()
                if (num<0.5){
                    ball.vx=-size
                }else{
                    ball.vx=size
                }
            }
        }else{
            ball.vx=0
            if(belowBall==-1){
                if(r<nbRows-1 && c<nbCols-1 && grid[r+1][c+1]==0){
                    ball.x=ball.x+size
                    ball.y=ball.y+size
                }else if(r<nbRows-1 && c>0 && grid[r+1][c-1]==0){
                    ball.x=ball.x-size
                    ball.y=ball.y+size
                }else{
                    grid[r][c]=-1
                }
                ball.vy=0
            }
        }
    }
}

function below(pR, pC){
    if(pR>=nbRows-1){
        return -1
    }
    return grid[pR+1][pC]
}

function drawGrid(){
    ctx.fillStyle="rgb(51, 0, 51)"
    for(let i=0; i<nbRows; i++){
        for(let j=0; j<nbCols; j++){
            let value=grid[i][j]
            let x=j*size
            let y=i*size
            if(value==1){
                ctx.beginPath()
                ctx.arc(x+size/2, y+size/2, size/2, 0, 2*Math.PI)
                ctx.fill()
            }else if(value==2){
                ctx.fillRect(x, y, size, size)
            }else if(value==3){
                ctx.fillRect(x, y, size/2, size)
            }else if(value==4){
                ctx.fillRect(x+size/2, y, size/2, size)
            }
        }
    }
}

function drawBalls(){
    for(let i=0; i<listBalls.length; i++){
        let ball=listBalls[i]
        ctx.fillStyle="rgb("+ ball.color[0]+ ", "+ ball.color[1]+ ", "+ ball.color[2]+ ")"
        ctx.beginPath()
        ctx.arc(ball.x, ball.y, ball.r, 0, 2*Math.PI)
        ctx.fill()
    }
}

function draw(){
    ctx.fillStyle="rgb(0, 0, 0)"
    ctx.fillRect(0, 0, canW, canH)
    //ctx.save()
    //ctx.scale(1/2, 1/2)
    drawGrid()
    drawBalls()
    //ctx.restore()
}

function gameLoop(time){
    if(oldTime!=0){
        dt=(time-oldTime)/1000
    }else {
        dt=0
    }
    update(dt)
    draw()
    oldTime=time
    requestAnimationFrame(gameLoop)
}

File: galton_board_experiment/C/src/simulation.c
#include <stdio.h>
#include <math.h>
#include "simulation.h"

int centerC  = floor(NB_COLS/2); 
int grid[NB_ROWS][NB_COLS]; 
int countBall =0; 
ball_t* listBalls = NULL; 
float timerIntervalAdd = DURATION_INTERVAL_ADD; 
float timerIntervalMove = DURATION_INTERVAL_MOVE; 

int int_random(int n){
    return rand() % n; 
}

void initGrid(void){
	for(int i = 0; i < NB_ROWS; i++){
		for(int j = 0; j < NB_COLS; j++){
			if (i < START_NB_ROWS - 1){
				if (j == centerC-1){
					grid[i][j] = 3;
				}else if(j == centerC + 1){
					grid[i][j] = 4;
				}else{
					grid[i][j] = 0;
				}
			}else if (i < START_NB_ROWS){
				grid[i][j] = 0;
			}else if (i < START_NB_ROWS + 2 * NB_STEPS){
				if ((i - START_NB_ROWS) % 4 == 0){
					if(j % 4 == 0){
						grid[i][j] = 1;
					}else{
						grid[i][j] = 0;
					}
				}else if ((i - START_NB_ROWS) %4 == 1 || (i - START_NB_ROWS) % 4 == 3){
					grid[i][j] = 0;
				}else if ((i - START_NB_ROWS) % 4 == 2){
					if (j % 4 == 2){
						grid[i][j] = 1;
					}else{
						grid[i][j] = 0;
					}
				}
			}else{
				if(j % 4 == 2){
					grid[i][j] = 2;
				}else{
					grid[i][j] = 0;
				}
			}
		}    
	}
}

void addBall(void){
	ball_t ball; 
	ball.x = centerC * SIZE; 
	ball.y = 0; 
	ball.r = SIZE/2 -1; 
	ball.vx = 0; 
	ball.vy = SIZE; 
	listBalls[countBall++] = ball; 
}

void restart(void){
	timerIntervalAdd = DURATION_INTERVAL_ADD; 
	timerIntervalMove = DURATION_INTERVAL_MOVE;
	countBall = 0; 
	initGrid();  
	addBall(); 
}

int below(int r, int c){
	if (r >= NB_ROWS - 1){
		return -1;
	}
	return grid[r+1][c];
}

void updateBalls(void){
    for(int i=0; i<countBall; i++){
        ball_t* ball = &listBalls[i];
        ball->x = ball->x + ball->vx;
        ball->y = ball->y + ball->vy;
    	int r = floor(ball->y / SIZE);
        int c = floor(ball->x / SIZE);
		//printf("r and c %d %d %d %d %d\n", r, c, ball->x, ball->y, i);
        int belowBall = below(r, c);
		//printf("belowBall %d \n", belowBall); 
        if(r < START_NB_ROWS + 2 * NB_STEPS - 1){
            if(belowBall == 1){
                int num = int_random(2); 
                if (num == 0){
                    ball->vx = -SIZE;
                }else{
                    ball->vx = SIZE;
                }
            }
        }else{
            ball->vx = 0;
            if(belowBall == -1){
                if(r < NB_ROWS-1 && c < NB_COLS-1 && grid[r+1][c+1] == 0){
                    ball->x = ball->x + SIZE;
                    ball->y = ball->y + SIZE;
                }else if(r < NB_ROWS - 1 && c > 0 && grid[r+1][c-1] == 0){
                    ball->x = ball->x - SIZE;
                    ball->y = ball->y + SIZE;
                }else{
                    grid[r][c] = -1;
                }
                ball->vy = 0; 
            }
        }
    }
}

void updateSimulation(float dt){
	timerIntervalAdd -= dt;
    timerIntervalMove -= dt;
	if (timerIntervalMove <= 0){
        updateBalls();
        timerIntervalMove = DURATION_INTERVAL_MOVE;
    } 
    if (countBall < TOTAL_BALLS){
        if (timerIntervalAdd <= 0){
            addBall();
            timerIntervalAdd = DURATION_INTERVAL_ADD;
        }
    } 
}

void drawGrid(uint32_t color) {
	for(int i = 0; i < NB_ROWS; i++){
		for(int j = 0; j < NB_COLS; j++){
			int value = grid[i][j];
            int x = j * SIZE;
            int y = i * SIZE;
			if(value==1){
				fillRect(x, y , SIZE, SIZE, color); 
			}else if (value==2){
				fillRect(x, y, SIZE, SIZE, color); 
			}else if (value==3){
				fillRect(x, y, SIZE/2, SIZE, color); 
			}else if (value==4){
				fillRect(x + SIZE/2, y, SIZE/2, SIZE, color); 
			}
		}
	}
}

void drawBalls(uint32_t color) {
	for(int i = 0; i < countBall; i++){
		ball_t ball = listBalls[i];
		fillRect(ball.x, ball.y, 2*ball.r, 2*ball.r, color); 
	}
}